File: processor/src/main/java/com/qualityminds/lazyval/processor/codegen/JpaGenerator.java
package com.qualityminds.lazyval.processor.codegen;

import com.palantir.javapoet.*;
import com.qualityminds.lazyval.processor.spi.FilePerTypeGenerator;
import com.qualityminds.lazyval.processor.spi.GeneratorResult;
import com.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import java.util.Collection;
import java.util.List;
import java.util.Set;

// must only be public for ServiceLoader, but it is not part of the API
@SuppressWarnings("doclint:accessibility,missing")
// tag::docu[]
public class JpaGenerator implements FilePerTypeGenerator {

    private static final String GENERATOR_ID = "jpa";
    private static final String OPTION_GENERATED_PACKAGE = "lazyval.jpa.generatedPackage";

    @Override
    public String generatorId() {
        return GENERATOR_ID;
    }

    @Override
    public Collection<String> requiredClasspath() {
        return List.of("jakarta.persistence.AttributeConverter");
    }

    @Override
    public Set<String> supportedOptions() {
        return Set.of(OPTION_GENERATED_PACKAGE);
    }

    @Override
    public GeneratorResult generateFilePerType(ValidatedGeneratorElement validElement, Settings userSettings){
        // end::docu[]
        TypeMirror type = validElement.element().asType();
        var wrappedType = validElement.wrappedType();
        TypeName wrappedTypeName;
        if (wrappedType.isPrimitive()) {
            // Box primitive types for JPA generics
            wrappedTypeName = TypeName.get(wrappedType.typeMirror()).box();
        } else {
            wrappedTypeName = TypeName.get(wrappedType.typeMirror());
        }

        var parameterName = "typeMirror";
        // Jpa Converter needs to check for null due to boxing primitive types
        final MethodSpec convertToDatabaseColumn = MethodSpec.methodBuilder("convertToDatabaseColumn")
                    .addModifiers(Modifier.PUBLIC)
                    .returns(wrappedTypeName)
                    .addParameter(TypeName.get(type), parameterName)
                    .beginControlFlow("if(%s == null)".formatted(parameterName))
                    .addStatement("return null")
                    .endControlFlow()
                    .addStatement(String.format("return %s.%s", parameterName, validElement.accessor()))
                    .build();

        parameterName = "dbValue";
        final MethodSpec convertToEntityAttribute = MethodSpec.methodBuilder("convertToEntityAttribute")
                    .addModifiers(Modifier.PUBLIC)
                    .returns(TypeName.get(type))
                    .addParameter(wrappedTypeName, parameterName)
                    .beginControlFlow("if(%s == null)".formatted(parameterName))
                    .addStatement("return null")
                    .endControlFlow()
                    .addStatement("return $L", validElement.objectCreation(parameterName))
                    .build();

        TypeSpec jpaConverter = TypeSpec.classBuilder(validElement.typeName() + "AttributeConverter")
                .addAnnotation(AnnotationSpec.builder(ClassName.get("jakarta.persistence", "Converter"))
                        .addMember("autoApply", "$L", "true")
                        .build())
                .addSuperinterface(ParameterizedTypeName.get(
                        ClassName.get("jakarta.persistence", "AttributeConverter"),
                        TypeName.get(type),
                        wrappedTypeName))
                .addModifiers(Modifier.PUBLIC)
                .addMethod(convertToDatabaseColumn)
                .addMethod(convertToEntityAttribute)
                .build();

        String jpaConverterPackage = userSettings.get(OPTION_GENERATED_PACKAGE)
                .orElse(String.format("%s.boundary.persistence", extractRootPackage(validElement.element())));
        if(jpaConverterPackage.charAt(0) == '.'){
            jpaConverterPackage = jpaConverterPackage.substring(1);
        }
        final JavaFile javaFile = JavaFile.builder(jpaConverterPackage, jpaConverter).build();
        return new GeneratorResult.Java(
                new GeneratorResult.Metadata(javaFile.packageName(), javaFile.typeSpec().name()),
                javaFile.toString());
    }
}

File: ksp/src/main/kotlin/de/qualityminds/lazyval/ksp/codegen/JpaGenerator.kt
package de.qualityminds.lazyval.ksp.codegen

import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import de.qualityminds.lazyval.ksp.ValidatedKspGeneratorElement
import de.qualityminds.lazyval.ksp.spi.FilePerTypeGenerator
import de.qualityminds.lazyval.ksp.spi.GeneratorResult
import de.qualityminds.lazyval.ksp.spi.SpiGenerator

// tag::docu[]
class JpaGenerator : FilePerTypeGenerator {

    companion object {
        const val OPTION_GENERATED_PACKAGE = "lazyval.jpa.generatedPackage"
    }

    override fun generatorId(): String = "jpa"

    override fun requiredClasspath(): Collection<String> = listOf("jakarta.persistence.AttributeConverter")

    override fun generateFilePerType(
        validatedElement: ValidatedKspGeneratorElement,
        userSettings: SpiGenerator.Settings
    ): GeneratorResult {
        // end::docu[]
        val element = validatedElement.element
        val wrappedType = validatedElement.wrappedType
        val lazyvalTypeName = element.toClassName()

        // Handle primitive boxing for JPA generics
        val wrappedTypeName = if (wrappedType.isBoxedPrimitive() || wrappedType.isPrimitive()) {
            wrappedType.toTypeName().copy(nullable = false)
        } else {
            wrappedType.toTypeName()
        }

        val converterClassName = "${element.simpleName.asString()}AttributeConverter"

        // Build convertToDatabaseColumn method - use the Kotlin accessor method name
        val convertToDatabaseColumn = FunSpec.builder("convertToDatabaseColumn")
            .addModifiers(KModifier.OVERRIDE)
            .returns(wrappedTypeName.copy(nullable = true))
            .addParameter("type", lazyvalTypeName.copy(nullable = true))
            .apply {
                addStatement("return type?.${validatedElement.kotlinAccessorMethodName}")
            }
            .build()

        // Store factory method in local variable to avoid smart cast issues
        val factoryMethod = validatedElement.factoryMethod
        val objectCreation = if (factoryMethod != null) {
            "${lazyvalTypeName.simpleName}.${factoryMethod.simpleName.asString()}(dbValue)"
        } else {
            "${lazyvalTypeName.simpleName}(dbValue)"
        }

        // Build convertToEntityAttribute method
        val convertToEntityAttribute = FunSpec.builder("convertToEntityAttribute")
            .addModifiers(KModifier.OVERRIDE)
            .returns(lazyvalTypeName.copy(nullable = true))
            .addParameter("dbValue", wrappedTypeName.copy(nullable = true))
            .apply {
                addStatement("return dbValue?.let { $objectCreation }")
            }
            .build()

        val converterClass = TypeSpec.classBuilder(converterClassName)
            .addAnnotation(
                AnnotationSpec.builder(ClassName("jakarta.persistence", "Converter"))
                    .addMember("autoApply = true")
                    .build()
            )
            .addSuperinterface(
                ClassName("jakarta.persistence", "AttributeConverter")
                    .parameterizedBy(
                        lazyvalTypeName.copy(nullable = true),
                        wrappedTypeName.copy(nullable = true)
                    )
            )
            .addFunction(convertToDatabaseColumn)
            .addFunction(convertToEntityAttribute)
            .build()

        // Determine package
        val packageName = userSettings.options[OPTION_GENERATED_PACKAGE]
            ?: "${extractRootPackage(element)}.boundary.persistence"

        val fileSpec = FileSpec.builder(packageName, converterClassName).addType(converterClass).build()

        return GeneratorResult.Kotlin(
            GeneratorResult.Metadata(packageName, converterClassName),
            fileSpec.toString())

    }
}